test/upgrade-matrix/mzcompose.py

```diff
@@ -0,0 +1,326 @@
+    BeginUpgradeScenario,
+    BeginVersion,
+    ChecksInitialize,
+    ChecksManipulate1,
+    ChecksManipulate2,
+    ChecksValidate,
+    EndUpgradeScenario,
+    EndVersion,
+    Node,
+)
+
+from materialize.checks.actions import Action
+from materialize.checks.all_checks import *  # noqa: F401 F403
+from materialize.checks.checks import Check
+from materialize.checks.executors import MzcomposeExecutor
+from materialize.checks.scenarios import *  # noqa: F401 F403
+from materialize.checks.scenarios import Scenario
+from materialize.mzcompose import Composition, WorkflowArgumentParser
+from materialize.mzcompose.services import (
+    Cockroach,
+    Debezium,
+    Materialized,
+    Postgres,
+    Redpanda,
+)
+from materialize.mzcompose.services import Testdrive as TestdriveService
+from materialize.util import MzVersion, released_materialize_versions
+
+# All released Materialize versions, in order from most to least recent.
+all_versions = released_materialize_versions()
+
+SERVICES = [
+    Cockroach(setup_materialize=True),
+    Postgres(),
+    Redpanda(auto_create_topics=True),
+    Debezium(redpanda=True),
+    Materialized(),  # Overriden inside Platform Checks
+    TestdriveService(
+        default_timeout="300s",
+        no_reset=True,
+        seed=1,
+        entrypoint_extra=[
+            "--var=replicas=1",
+            f"--var=default-replica-size={Materialized.Size.DEFAULT_SIZE}-{Materialized.Size.DEFAULT_SIZE}",
+            f"--var=default-storage-size={Materialized.Size.DEFAULT_SIZE}",
+        ],
+    ),
+]
+
+
+def setup(c: Composition) -> None:
+    c.up("testdrive", persistent=True)
+    c.up("cockroach", "redpanda", "postgres", "debezium")
+
+
+def teardown(c: Composition) -> None:
+    c.down(destroy_volumes=True)
+
+
+def workflow_default(c: Composition, parser: WorkflowArgumentParser) -> None:
+    """Test sequences of upgrades across the entire possible matrix of upgrade scenarios.
+
+    Thie framework works by building a graph that contains the Mz versions and Checks actions
+    to perform as nodes, as well as the allowed upgrade paths and other required dependencies
+    (as edges). All valid simple paths within the graph are upgrade scenarios that need to be
+    exercised and we pick a subset of them to run.
+    """
+
+    c.silent = True
+    parser.add_argument(
+        "--check",
+        metavar="CHECK",
+        type=str,
+        action="append",
+        help="Check(s) to run.",
+        default=["PgCdc", "SinkUpsert"],
+    )
+
+    parser.add_argument(
+        "--dry-run",
+        action="store_true",
+        help="Just print the upgrade paths that will be tested.",
+    )
+
+    parser.add_argument(
+        "--min-version",
+        metavar="VERSION",
+        type=MzVersion.parse,
+        default=MzVersion.parse("0.39.0"),
+        help="Minimum Mz version to involve.",
+    )
+
+    parser.add_argument(
+        "--max-version-gap",
+        type=int,
+        default=1,
```

Comment:
Why just 1? I think 2-3 would be more interesting and still realistic.

Reply:
This is still under discussion. @benesch thinks that with careful orchestration of the upgrade process in production, we will never ever have gaps , even single versions, when upgrading actual customers ( so max-version-gap=0). I tend to disagree, because there may always be unexpected upgrade emergencies and such.

So I will be presenting a dedicated proposal as to what constitutes a valid upgrade scenario and what not, including the issue on gaps. In the meantime, I put "default=1", because it things are currently green with it and it is a less controversial setting as compared to 2 or more.